### PR TITLE
Hinted handoff bug

### PR DESCRIPTION
**Pull request by davecromberge**
The hinted handoff process is triggered once a vnode enters an inactive state. This happens when the current active state times out, for a period of anywhere between 60s and 120s. basho/riak_core#715

A metric_vnode updates it's current time on a 1s tick, which is used to determine if a metric has a valid timestamp. This has the side effect of prolonging the active state, preventing the vnode from becoming idle.

Instead, the timestamp is now recreated on every invocation to the expiry function, allowing the tick message to be removed.

### CODE REVIEW

**Comment by davecromberge** [comment]
@Licenser I include your feedback in response to this PR:

> Every put executes valid_ts -> expiery, so effectively removing the timer would cause every put request run erlang:system_time while it is supposed to be not as bad as erlang:now() was I’m not entirely sure it’s good, if you got a few M metrics each second adding a few million calls to erlang:system_time seems like a lot.

> One thing I could see is setting the tick timeout to something like 
https://github.com/basho/riak_core/blob/develop/src/riak_core_vnode.erl#L244

> + ?TIMEOUT

> that would ensure that we got a periodic update on the timestamp but it seems it would be a bit high > (1min+).

> Doing a cheap, not 100% representative benchmark:

> 2> timer:tc(fun () -> [erlang:system_time() || _ <- lists:seq(0, 1000000)], ok end).
{1575953,ok}

> having 1M erlang:system_time() calls on my mac makes for ~1.5s, DDB could handle ~ 2M  on a single node (IOBound) so this would be a serious impact on performance.

My earlier attempt was to retain the tick functionality and adjust the interval beyond that in the riak_core_vnode ?TIMEOUT value.  This did prove successful, but I am sure that this affects the calculation that the metric vnode performs regarding the Bucket lifetime threshold:
https://github.com/dalmatinerdb/dalmatinerdb/blob/test/apps/metric_vnode/src/metric_vnode.erl#L498

To my understanding this seems pretty reliant on the 1s interval.  I don't know the system well enough to determine the impact of making the expiry calculation less accurate.  It seems like a compromise between correctness and performance, given this PR.  I needed the handoffs, and since we are using the infinite lifetime at this stage the performance impact of the change is a non-factor.  However, I would like to propose the changes nevertheless, to debate and explore alternatives.  It's also good to have a record of the conversation :smile:

**Comment by Licenser** [comment]
Your patch will still introduce the overhead of running system_time for every single put request even when you do not expire at all.

I am fairly certain this is the wrong approach, to solve the issue. The real solution here would be fixing the related riak_core issue, the second best I think would be bumping the timeout. it will 'only' mean that data vacuum will be delayed a few cycles.

**Comment by davecromberge** [comment]
Ok, I think that is an acceptable outcome.  Given that the impact of the Timestamp delay appears minimal, I can resubmit this with a new ?TICK value that exceeds the `inactive_timeout` interval.
I still don't see how the time would affect it for buckets with a lifetime set to infinity, as it would match on the first case:
```erlang
expiry(Bucket, State) ->
    case get_lifetime(Bucket, State) of
        {infinity, State1} ->
            {infinity, State1};
        {LT, State1} ->
            {Res, State2} = get_resolution(Bucket, State1),
            Now = erlang:system_time(milli_seconds),
            Exp = (Now - LT) div Res,
            {Exp, State2}
    end.


```

**Comment by Licenser** [comment]
You are right it won't, I was looking at a old version of your patch (https://github.com/dataloop/dalmatinerdb/commit/62ee0f4d78a028e8ace384b5dd53168b3053a42c) instead of the latest where the system_time is in the second branch.

Ultimately both are just workarounds, and with infinity buckets neither would have an impact on your system, so I think the longer tick times are a better overall compromise on not impacting the current behavior too much, after all the 'vacuum' command isn't executed on an instant.

Thought, the timestamp could be updated every time vacuum is triggered the vacuum impact would also be neglect able.

**Comment by davecromberge** [comment]
I have updated the PR to modify the timestamp during the vacuum cycle, and have verified that handoffs still work.  Is that what you had in mind?

File: apps/metric_vnode/src/metric_vnode.erl [hunk 1]
```diff
@@ -474,27 +467,40 @@ do_put(Bucket, Metric, Time, Value,
...
             {false, State1};
         {_, State1} ->
             {true, State1}
     end.
+
 %% Return the latest point we'd ever want to save. This is more strict
 %% then the expiration we do on the data but it is strong enough for
 %% our guarantee.
-expiery(Bucket, State = #state{now = Now}) ->
+expiry(Bucket, State) ->
```
Comment (Licenser) [inline apps/metric_vnode/src/metric_vnode.erl]: this function should remain unchanged since `now` is still (or agian) in the `state` record

**Comment by Licenser** [comment]
Added one inline comment but I think it's generally good, having the `timestamp` in the vacuum ensures that on data is ever written into the already deleted time range. Some unnecessary data might be written in edge cases [1] but think that is a tradeoff worth making to fix the handoffs.

[1] current code would never write in a time range that already is in the vacuum time range, this change would but never in a already deleted ranges.

File: apps/metric_vnode/src/metric_vnode.erl (hunk 1 above)
Comment (davecromberge) [inline apps/metric_vnode/src/metric_vnode.erl]: Oops! I forgot to add this back, good catch.

**Comment by davecromberge** [comment]
The tradeoff is unfortunate, but I agree, worth making for the handoffs as you said.  Glad you approve of the approach, thanks for all your feedback on this PR so far.

**Comment by Licenser** [comment]
looks great :) thanks mate!

**Comment by davecromberge** [comment]
Awesome!

**Pull request merged**